### ISSUE TITLE
[RFC] Main menu "Tools" and "More tools"

### ISSUE THREAD

**Issue by hius07** [koreader/koreader#9638]
**Purposes**
(1) Optimize Tools menu
(2) Simplify plugins integration

**Menu rearrangement**
(1) Remove "More tools" section from the "Tools" menu

(2) "Tools" section includes:
-"Plugins management"
-separator
-all (enabled) plugins in alphabetical order

(3) Add new (tenth) section "Advanced settings" to the "Settings" menu that includes:
-"Defaults"
-"Developer options"

**Code changes**
(1) filemanager_menu_order and reader_menu_order do not include elements under Tools section
(2) plugin developer does not take care of placing the plugin item into the main menu (unless the plugin is placed outside the "Tools" section)
(3) plugin must include `sorting_hint` with values `tools` or `tools_filemanager` or `tools_reader`
(4) MenuSorter takes care of collecting enabled plugins into the sorted Tools section based on the `sorting_hint` value

**Negative consequences**
(1) Disables user's sorting order of the plugins

**Comment by Frenzie** [comment]
> (2) "Tools" section includes:
-"Plugins management"
-separator
-all (enabled) plugins in alphabetical order

That's the exact opposite of the design principle. The *tools* (not plugins!) are sorted *logically*, not arbitrarily.

**Comment by hius07** [comment]
What is the current logic?
https://github.com/koreader/koreader/blob/d9eb6e971742810bb1ee545f80686a84b90c06d3/frontend/ui/elements/filemanager_menu_order.lua#L129-L135

**Comment by poire-z** [comment]
No strong opinion about either current vs. this proposal.

There might be a thing with alphabetical order: all the `Auto-something` plugins will be first, and `Reading statistics` (the most sexy for newcomers, and probably the most used called from the menu) will probably be on page 2.

> Add new (tenth) section "Advanced settings" to the "Settings" menu that includes:
-"Defaults"
-"Developer options"

What's the "Settings" menu . The gear ? (A bit sad to waste the 10th and last slot for this :)
Also, what's "Defaults" ? "Reset settings to defaults" found in Style/Font/Typo menu ?

**Comment by hius07** [comment]
"Plugins management" can be a SortWidget with checkboxes, to enable/disable plugins and sort the Tools menu.
The order can be saved in settings along to `plugins_disabled`.



> what's "Defaults" ?

Current "Advanced settings" that shows the MultiInputDialog titled "Defaults".

**Comment by Frenzie** [comment]
> What is the current logic?

I didn't say there's no room for improvement. ^_^ The logic is some combination of most useful to be quickly accessible or otherwise most important to less used combined with logical grouping. Hence why in tools the read timer comes first and in settings the light and night mode comes first for example. The fact that I barely use the light and never use night mode aren't relevant to the reasoning. ;-) The terminal emulator is far down because it's a lot like the developer options — most users probably won't use it regardless how nice it is.

More tools is presumably one of the less well-ordered menus internally. The main logic is that its content is explicitly in more tools and not in tools. I forget exactly why it exists though; I'm not fond of it.

**Comment by pazos** [comment]
I think that part of the "problem" is that we enable every single plugin on every single platform. And things are messy for new users.

I don't believe any long time KO user keeps all those plugins enabled, cluttering the UI, without any usage.

My 2 cents: instead of reordering tools make a table of *good enough for vanilla KO experience* plugins. Ship those enabled by default on new install.

99% of users will be happy with only `coverbrowser` enabled by default and little hints on the quickstart guide (and the full blown guide) that explains them how to enable plugins and what they do.

**Comment by NiLuJe** [comment]
@pazos: I'd tend to agree with that, and would possibly make this extremely explicit by *forcing* the QuickStart xit to go straight to the Plugin management window?

Or something equally annoying to make it very obvious you *need* to go there, soon.

**Comment by hius07** [comment]
That's one more reason to get rid of More tools submenu.

**Comment by Frenzie** [comment]
I'm not quite convinced by that argument. Many people ask me about Pocket because they like it in Nickel, and then I have to explain that Wallabag is pretty much the same thing. Adding another step by first having to find and enable the plugin will only make that worse.

As a counterexample, you could give the network/online/export/sync tools their own submenu together (calibre, sync, Wallabag, NewsDownloader, highlight export, QR, any others?). That's the positive version of the "more tools" menu that better embodies the menu principles as I see them. Note that this could also be a submenu in the main/hamburger menu.

I do sometimes wonder if the Perception Expander Plugin should be disabled by default because people enable it and forget or don't realize they did, but menu-wise I already relocated that outside of tools to a better location several years ago.

To counter my initial paragraph, disabling most plugins by default would make the plugin manager significantly easier to find. But in that case it should definitely look more like this:

<img src=https://user-images.githubusercontent.com/202757/195939493-267d2c64-662e-47d7-939d-a4d43e6c39e8.png width=45%> <img src=https://user-images.githubusercontent.com/202757/195939567-d2159dfc-2810-4ef4-8303-4fa11564814e.png width=45%>

**Comment by roygbyte** [comment]
> As a counterexample, you could give the network/online/export/sync tools their own submenu together (calibre, sync, Wallabag, NewsDownloader, highlight export, QR, any others?). That's the positive version of the "more tools" menu that better embodies the menu principles as I see them. Note that this could also be a submenu in the main/hamburger menu.

To me, this counterexample evinces a failure with the existing menu structure, which is its inability to distinguish between plugins that extend or tweak reading features or functionality, and plugins that enable an entirely new functionality. In other words, some of these plugins feels more like apps... they don't just add a feature, they add a new way of gathering content, interpreting data, running a terminal, and so on. I'd like to see a menu structure that reflects the type of engagement expected for the different types of plugins. Like, one could expect to launch Newsdownloader a few times a week, so it should be easier to find, perhaps situation in a context where its next to other plugins (Wallabag, Text Editor) that are used with a similar frequency. Plugins that are "set and forget" like "synchronize time" or "Tweak document settings" are elsewhere, in a context that is more like the Kate configuration plugins shown above. 

To recap, the issue for me is that there *is* a plugins menu, when what there should be instead is more tight integration between what the plugin is, and logically where a person would expect to find it. I recognize this might expand the scope of the original suggestion. But maybe there's a bigger issue here.

**Comment by Frenzie** [comment]
> To recap, the issue for me is that there is a plugins menu

Repeat after me: *there is no such thing as a plugins menu*.

**Comment by Frenzie** [comment]
It's the *tools* menu. The fact that they are plugins is a completely irrelevant technical detail.

**Comment by Frenzie** [comment]
> To recap, the issue for me is that there is a plugins menu, when what there should be instead is more tight integration between what the plugin is, and logically where a person would expect to find it. I recognize this might expand the scope of the original suggestion. But maybe there's a bigger issue here, so it warrants expanding focus.

No, that *is the design principle*.

**Comment by pazos** [comment]
I must insist: it is mostly a non problem for us contributors. My daily driver has only four things on the tools menu and that's because cloud storage is not a plugin yet :p.

If somebody really uses more than 4-5 tools on a regular basis I would suggest to focus on porting common actions to the dispatcher and forget about menus :)

**Comment by foundObjects** [comment]
> I think that part of the "problem" is that we enable every single plugin on every single platform. And things are messy for new users.
> 
> I don't believe any long time KO user keeps all those plugins enabled, cluttering the UI, without any usage.
> 
> My 2 cents: instead of reordering tools make a table of _good enough for vanilla KO experience_ plugins. Ship those enabled by default on new install.
> 
> 99% of users will be happy with only `coverbrowser` enabled by default and little hints on the quickstart guide (and the full blown guide) that explains them how to enable plugins and what they do.

User here, the first thing I do when I install KOReader on a new device is go disable a bunch of unused plugins to simplify menus. I like the idea of shipping with a small number of basic plugins enabled. I'd add Gestures to that bucket too.

**Comment by offset-torque** [comment]
> **Frenzie:** The logic is some combination of most useful to be quickly accessible or otherwise most important to less used combined with logical grouping. Hence why in tools the read timer comes first.

**Read Timer** is officially the most important plugin ?!?
Maybe it should start working like a real **read timer** instead of a kitchen **egg timer** to deserve its 1st place in the menu ;)

Here is a reminder: [Improvements to the Read Timer](https://github.com/koreader/koreader/issues/9216)

**Comment by Frenzie** [comment]
I'd love to do a @weijiuqiao style update of it. :-)

**Comment by weijiuqiao** [comment]
I'm flattered, though I'm not really sure what it means here this time :-)

**Comment by Frenzie** [comment]
Nothing really, I just wanted to call out the nice UI you made.

**Comment by lightmaster** [comment]
> @pazos: I'd tend to agree with that, and would possibly make this extremely explicit by _forcing_ the QuickStart exit to go straight to the Plugin management menu?
> 
> Or something equally annoying to make it very obvious you _need_ to go there, soon.

I've been using KOReader for a couple months now, and until I stumbled upon this thread, I had now idea you could disable plugins to clean up the Tools menu. Definitely think only enabling a subset of plugins by default, and then forcing all users to go to plugin manager menu so they can enable/disable what they want, and so they can even know that's possible.

**Comment by alexandrglm** [comment]
Hello,
a KOReader user here, on a Kobo Clara Colour.

First of all, I would like to express my gratitude for all the genuine functionality that KOReader provides, and for the support offered.

I must say that it should not be overly complicated to have access to the Terminal without having to navigate through four layers of menus beforehand.

It is a very, very, very, when I say "Very" it is really very, useful plugin, and there is no humanly possible way to load KOReader directly into said plugin (which would even help to create a direct button for those of us who use NickelMenu). Nor does it make sense to "hide" access to the terminal so deeply.

I join the previous similar requests; I am sure we can find a way to improve this.

Thanks,
Alexander

**Comment by mergen3107** [comment]
Well, KO**Reader** is a utility software first and foremost :D 

You can also assign TE to a gesture, or make it fire up on KOReader start (I think) with Profile auto exec. 

At the same time I agree either way this thread that there should be more streamlined menu organization. Most of the built-in plugins were designed on the spot somewhere between 2012-2104 and took their places in the top hierarchy. Then newer (and more useful) plugins show up to an already over crowded menu lists :D

**Comment by alexandrglm** [comment]
> Well, KO**Reader** is a utility software first and foremost :D

Yes, I admit it extends the original purpose of KO**reader**, *ha ha*!  
But even so, I think plugins like Terminal are just as widely used as File Manager or Reader functions.

> You can also assign TE to a gesture, or make it fire up on KOReader start (I think) with Profile auto exec.

Done, and useful, but not enough (*ha ha* again).

> At the same time I agree either way this thread that there should be more streamlined menu organization. Most of the built-in plugins were designed on the spot somewhere between 2012-2104 and took their places in the top hierarchy. Then newer (and more useful) plugins show up to an already over crowded menu lists :D

Please, consider this, or consider a way to get a new "Custom menu" with users' selected options or plugins

Greetings!

**Comment by offset-torque** [comment]
> a way to get a new "Custom menu" with users' selected options or plugins

We already have exactly this @alexandrglm : `Quick Menu`
I don't know if you are aware of this but **Terminal emulator** can be added  to your QM (which is much faster than going into menu system btw). This should solve your problem.

---

But if the request is for other plugins too, maybe we can add a new `Plugins` category to **Gesture Manager** which will list the currently installed Plugins. When added to a QM, these QM items might do one of these:
1. Open the plugin's menu as if it is clicked from the Tools menu
2. Or directly render the plugin's menu in a dialog

I don't know if number **2** is possible (or a good idea), but I guess **1** should work, since we have a "walk me there" feature in the **Menu search** which emulates user navigation. What do you say @hius07 ?

Benefit of this QM approach is, we can provide quick access to plugins without touching the main menu system. Users can build their own Plugins menu and we already have the ui for this as QM/Gesture Manager selection procedure.

**Comment by hius07** [comment]
Currently the Terminal can be started:
-from the upper menu (4 taps)
-with a gesture (1 gesture)
-from custom QuickMenu (1 gesture + 1 tap)
-at KOReader start (none)
Isn't this enough?

**Comment by Frenzie** [comment]
I think it could be interesting to be able to start some of these with a command line argument as well. (In a generic way with profiles or dispatcher.) Very low priority for me though. :)

**Comment by shuvashish76** [comment]
> maybe we can add a new `Plugins` category to Gesture Manager which will list the currently **installed** Plugins.

Few observations:
Installed plugins = User plugins ? There are various category of plugins and plugin-devs are free to add to whatever category they think is appropriate for their **plugin actions**.

> When added to a QM, these QM items might do one of these:
> 1. Open the plugin's menu as if it is clicked from the Tools menu

But not all plugins are in [Tools menu](https://koreader.rocks/user_guide/#L2-topexplained), and if I understand correctly this feature overlap with "Menu Search" which I think is good enough for users.

- Alternative suggestion: Quick search > Search results > When it shows "Cancel/Open/Walk_me_there" - add another option "Add as Gesture action"

Not sure if it's a good idea because there can be duplication of Gesture actions that are already available in the list.

> 2. Or directly render the plugin's menu in a dialog.
> Benefit of this QM approach is, we can provide quick access to plugins without touching the main menu system.

I think I've seen few plugins without having their own menu, they just start in background on start or with some condition or via gesture-action.

**Comment by offset-torque** [comment]
Good points @shuvashish76 

To clarify, I am not personally requesting this. I only use Export plugin and that one is already in the dispatcher.

I just wanted to remind that if there is a big demand for this kind of functionality, repurposing the QM workflow might be an option too. Especially considering that discussions for changing the main menu system is (expectedly) heated and exhausting.

**Comment by alexandrglm** [comment]
> @offset-torque said:
>
> We already have exactly this [@alexandrglm](https://github.com/alexandrglm) : `Quick Menu` I don't know if you are aware of this but **Terminal emulator** can be added to your QM (which is much faster than going into menu system btw). This should solve your problem.
> 

Cheers, mate. Now, thanks to the Quick Menu we all *obviously* knew existed, I’ve got direct access to a useless Terminal that refuses to stop rendering the QM over the top:
<img width="1072" height="1448" alt="Image" src="https://github.com/user-attachments/assets/abfc748e-1d61-47b2-8f3a-cbbbd3a3780e" />

> @Frenzie said:
>
> I think it could be interesting to be able to start some of these with a command line argument as w

Right, surely the plugin launcher could do with a review so we can run them directly from shell (or commnands via nickel menu, etc) ... a kind of direct call combining the main KOreader startup call with the plugin itself.

**Comment by mergen3107** [comment]
> Quick Menu we all obviously knew existed

That means you never discovered the User Guide? :D 

http://koreader.rocks/user_guide/#L1-qmandprofiles

**Comment by alexandrglm** [comment]
> @mergen3107 said:
>
> > Quick Menu we all obviously knew existed
> 
> That means you never discovered the User Guide? :D
> 
> http://koreader.rocks/user_guide/#L1-qmandprofiles

Nice try, dude. Means it's not working properly, which is what I'm trying to show with the image, it's not about reading mandocs. (If someone is going to offer such a bloody ooooobvious solution, they ought to test whether it actually works first ...in this case, any plugin launched via the Quick Menu ends up broken, as the UI never stops rendering the Quick Menu over plugins).

And while I'm here contributing, isn't this actually the most ideal spot for a proper Quick Menu?

<img width="1072" height="1448" alt="Image" src="https://github.com/user-attachments/assets/d2260d53-e373-4306-a63d-63219f6312c1" />

**Comment by mergen3107** [comment]
> they ought to test whether it actually works first

No one owes you a thing here, mate.

Did you read all the options in the guide? One of them is keeping the quick menu open after use (a checkbox). If this doesn't work, then open a bug report ticket.

**Comment by offset-torque** [comment]
> If someone is going to offer such a bloody ooooobvious solution, they ought to test whether it actually works first ...

Yeah I tested before and it works. After this overly dramatic reply I tested again now, and it still works. I launch Terminal Emulator from QM and menu disappears.

Check the `Keep QuickMenu open` setting in your QM. It exactly behaves like this when enabled, as the name says. If it is not checked, you can open an issue about it.

**Comment by hius07** [comment]
> useless Terminal that refuses to stop rendering the QM over the top

Uncheck "Keep QuickMenu open".

**Comment by mergen3107** [comment]
Haha here you go, three exact same replies a minute :D

**Comment by alexandrglm** [comment]
Fair cop, lads! Guilty as charged on the drama. My bad, dudes.

But I came here to discuss two main ideas:

1. **Direct CLI / Shell calls for plugins** (a way to launch specific KOreader plugins or functions directly via shell)

2. **A better home for the Quick Menu** (repositioning/re-integrating the Quick Menu into the main top menu bar, which feels like a much more natural and accessible spot for it)

Sorry for the earful!

> @mergen3107 said:
> > they ought to test whether it actually works first
> 
> No one owes you a thing here, mate.
> 
> Did you read all the options in the guide? One of them is keeping the quick menu open after use (a checkbox). If this doesn't work, then open a bug report ticket.

Off-topic, but that's quite an aggressive way to collaborate, dude.

**Comment by Frenzie** [comment]
The second isn't as obvious as it looks on your screenshot since devices don't all have the same aspect ratio. But folks are welcome to look into it to see if it can be elegantly done.

**Comment by mergen3107** [comment]
Yeah, second item needs a bit of overhaul. I partly support this idea because currently it is an adventure to set up a quick menu. Main struggle is that you need to assign it to a gesture. 

So once you setup a whole QM to one gesture, it is not too easy to move it all to another gesture. 

So this might be a good start. Make one common QM and place somewhere visibly reachable, as suggested as one example. 

Maybe also "long press to edit QM". 

Then make a dispatcher action to make it assignable to any gesture. 

Keep existing QM options as is (in gestures).

**Comment by hius07** [comment]
What is the benefit of calling a QM by two gestures (open the upper menu + tap)?
I can call any QM by a single gesture.

**Comment by mergen3107** [comment]
This is not only about how to quickly open a QM
Making one common QM (which is also more visible in the UI) makes it easier to re-assign it to any other gesture.
Currently you either need to setup all QM actions again inside another gesture or edit gestures file.

**Comment by hius07** [comment]
> Currently you either need to setup all QM actions again inside another gesture or edit gestures file.

I'm afraid I don't follow this.

**Comment by mergen3107** [comment]
For example:
1) I choose lower right corner tap for my QM
2) I select 10 actions
3) Select "Show as QM" in that gesture's top level. 
4) I use QM and realize I would like to have the same QM with 10 actions in a swipe gesture instead. 
5) So I open a swipe gesture, and do steps 1-3 manually again, then remove all actions from lower right corner type. 

In proposed common QM, steps 4-5 are replaced with:
4) Select 1 item (common QM) as action in new gesture
5) Unselect 1 item (the same common QM) from old gesture

**Comment by offset-torque** [comment]
I made a quick mockup of a zero friction Quick Menu.
It doesn't need further explanation or user guide.
Of course it has limitations but from the users' pov it can't get more intuitive than this.
Is it doable and/or is it a good idea I don't know. Just sharing as usual in case it might be:

<img width="1064" height="208" alt="Image" src="https://github.com/user-attachments/assets/eb142fbf-6e3c-498f-8dfb-e0119890c3dc" />

**NOTE:** I am not suggesting as a replacement of the current system.

**Comment by hius07** [comment]
> For example:

Just create a profile instead of a multi-action gesture in step 1.

**Comment by Frenzie** [comment]
I think that's meant to be read from the pov of an inexperienced user. :-)

**Comment by offset-torque** [comment]
> I think that's meant to be read from the pov of an inexperienced user. :-)

Definitely. I imagined a new user who has just installed KOReader and looking around.
This QM variant is discoverable and doesn't need documentation to set it up.
Power users can continue using the current system if they need more than one menu or nested menus for example.

**Comment by mergen3107** [comment]
> Just create a profile instead of a multi-action gesture in step 1.

Thanks, gotta try this :D 
My mind is too technical at this point to recognize easier solutions

@offset-torque mockup looks great!

**Comment by shuvashish76** [comment]
> in this case, any plugin launched via the Quick Menu ends up broken, as the **UI never stops rendering the Quick Menu over plugins**

If I understand your intention correctly then I reported about it at #15717
